### ISSUE TITLE
JIT: Model `Int64/UInt64 -> Single` casts without intermediate cast to `Double`

### ISSUE THREAD

**Issue by amanasifkhalid** [dotnet/runtime#106646]
Follow-up to #106419 ([discussion](https://github.com/dotnet/runtime/pull/106419#discussion_r1718839680)). When casting, for example, a `ulong` to a `float`, Roslyn emits the following IL:
```
conv.r.un
conv.r4
```
Which is then imported by the JIT as a `ulong -> float -> double` cast sequence. Casting in two steps can produce slightly different results from casting directly to `float`, which is also why it isn't always correct to just morph the above IR into `ulong -> float`. Ideally, we would differentiate between `ulong -> float` and `ulong -> double -> float` during importation, and create IR that accurately models each pattern.

Similarly, constant folding should model each pattern correctly. In other words, to model `ulong -> float`, we should cast the constant directly to `float`, and rely on MSVC/Clang/GCC/etc. to emit the correct sequence.

cc @dotnet/jit-contrib

**Comment by jakobbotsch** [comment]
> Ideally, we would differentiate between `ulong -> float` and `ulong -> double -> float` during importation, and create IR that accurately models each pattern.

In my opinion, as I mentioned in the discussion, the ideal fix is that Roslyn produces IL that implements `ulong -> float` conversions correctly. If we try to fix this in the JIT we end up with a bunch of odd possible behavior around how IL is handled.

**Comment by amanasifkhalid** [comment]
> In my opinion, as I mentioned in the discussion, the ideal fix is that Roslyn produces IL that implements ulong -> float conversions correctly. If we try to fix this in the JIT we end up with a bunch of odd possible behavior around how IL is handled.

I agree that that would simplify the JIT side. I'm guessing this will have to be a .NET 10 item if we want a fix in Roslyn, right? Should we open an issue in their repo?

**Comment by tannergooding** [comment]
> the ideal fix is that Roslyn produces IL that implements ulong -> float conversions correctly

I don't think any scenario where Roslyn emits new IL patterns would be viable and will likely cause large performance regressions.

Exposing a new intrinsic API (`RuntimeHelpers.ConvertToDouble(ulong)` and `RuntimeHelpers.ConvertToSingle(ulong)` for example) would be viable and ensure the JIT can still recognize the intent and optimize, but that still causes issues for existing code/libraries not matching the intent the user authored.

Since we are already in a world where we have oddities here based on what IL is present, whether optimizations are enabled or disabled, and what the underlying hardware supports instruction set wise. I don't think normalizing that to ensure that `conv.r.un; conv.r4` does the "right" thing is going to hurt that. I rather think that it will fix a bug for the vast majority of code out there (certainly anything currently being produced by Roslyn or F#) and only leave potential ambiguities for what is effectively manually written IL (which can be left as UB, which is functionally how it is already today). We could normalize such manually written IL as well, if that were desired (such as handling `conv.r.un; ...; conv.r4` with a gap in the IL, but no extra stores/uses could be handled for example as we could recognize the tree shape `CONV_R4(CONV_R_UN)` rather than only recognizing the explicit sequence).

**Comment by tannergooding** [comment]
> The compilers are emitting conv.r.un; conv.r4, which is not equivalent when the internal representation used for type F is larger than float/double

@jakobbotsch, this notably isn't about the internal representation being larger than float/double, that's never the case for RyuJIT for Arm64 or x64 (and shouldn't be the case for x86 since we use SSE/SSE2 and directly configure the x87 FPU for 64-bit rounding mode).

This is simply that `ulong -> double -> float` produces different from `ulong -> float` for some inputs (such as `16105307124325679103`) due to double rounding. There exists no singular IL instruction for doing `ulong -> float` (nor technically one for `ulong -> double`), hence the need for either recognizing `conv.r.un; conv.r4/conv.r8` as meaning `ulong -> float/double` rather than `ulong -> F -> float/double` or some new intrinssic API (such as `RuntimeHelpers.ConvertUInt64ToSingle/Double`) or some other suggested alternative.

**Comment by jakobbotsch** [comment]
> @jakobbotsch, this notably isn't about the internal representation being larger than float/double, that's never the case for RyuJIT for Arm64 or x64 (and shouldn't be the case for x86 since we use SSE/SSE2 and directly configure the x87 FPU for 64-bit rounding mode).
> 
> This is simply that `ulong -> double -> float` produces different from `ulong -> float` for some inputs (such as `16105307124325679103`) due to double rounding. There exists no singular IL instruction for doing `ulong -> float` (nor technically one for `ulong -> double`), hence the need for either recognizing `conv.r.un; conv.r4/conv.r8` as meaning `ulong -> float/double` rather than `ulong -> F -> float/double` or some new intrinssic API (such as `RuntimeHelpers.ConvertUInt64ToSingle/Double`) or some other suggested alternative.

I meant that `conv.r.un; conv.r8` is also not guaranteed to be equivalent to a `ulong -> double` conversion since the target IL compiler may use a larger representation than `double` for type `F`. I know that this is not the case for RyuJIT, so `conv.r.un; conv.r8` will be a proper `uklong -> double` conversion if RyuJIT happens to be compiling the IL.

**Comment by jaredpar** [comment]
> whether we should recognize conv.r.un; conv.r4/conv.r8 sequences specially and redefine their behavior, whether we should try providing intrinsics that the compilers can use, or maybe something entirely different.

In general I prefer having well defined intrinsics the compiler can use vs. trying to infer meaning from a series of op codes that somewhat deviates from the literal meaning of the op codes.

**Comment by dsyme** [comment]
cc @vzarytovskii @0101 @T-Gro

**Comment by vzarytovskii** [comment]
> cc @vzarytovskii @0101 @T-Gro 

Thanks. I'll see what do we emit after I'm back.

**Comment by vzarytovskii** [comment]
@jakobbotsch out of curiosity, what's the F# code you were looking at? If its something along the lines of `let convert (a: uint64) : float = float a`, then it should technically be straightforward for us to change, if we decide its the right thing to do.

**Comment by T-Gro** [comment]
> @jakobbotsch out of curiosity, what's the F# code you were looking at? If its something along the lines of `let convert (a: uint64) : float = float a`, then it should technically be straightforward for us to change, if we decide its the right thing to do.

The instruction sequence  is what gets emitted for the FSharp.Core's `float32`  function.
The same pattern is used for more input types:

```fsharp
             when ^T : uint64    = (# "conv.r.un conv.r4" value  : float32 #)
             when ^T : uint32    = (# "conv.r.un conv.r4" value  : float32 #)
             when ^T : uint16    = (# "conv.r.un conv.r4" value  : float32 #)
             when ^T : char      = (# "conv.r.un conv.r4" value  : float32 #)
             when ^T : unativeint = (# "conv.r.un conv.r4" value  : float32 #)
             when ^T : byte     = (# "conv.r.un conv.r4" value  : float32 #)
```

https://github.com/dotnet/fsharp/blob/bd9e200b5c0d177cc4a4155b59b8f52fbd4487c5/src/FSharp.Core/prim-types.fs#L5057-L5062

**Comment by saucecontrol** [comment]
Linking recent conversation on this topic: https://github.com/dotnet/runtime/pull/113930/files#r2017632169

In particular, this justification for how treating `conv.r.un, conv.r4` as a direct `ulong->float` cast can be legal by spec.

> It also fits with the concept of `F` in that we'd be treating it as if the conversion is to a "higher precision floating-point type" such that there would be no loss of data (and in fact for an 80-bit float `ulong->fp80->float` is the same as `ulong->float` since the `fp80` has a 64 bit significand

And this comment about the current optimization in `fgMorphExpandCast`.

> We definitely shouldn't be changing all `long->double->float` into `long->float` or all `ulong->double->float` into `ulong->float`. This would be bad and value changing in an undesirable way, we should really preserve it except for the specific conv.r.un, conv.r4 pattern and just document that tooling should respect that, which most already does and will continue to do.

**Comment by jakobbotsch** [comment]
> In particular, this justification for how treating `conv.r.un, conv.r4` as a direct `ulong->float` cast can be legal by spec.
> 

I think it would be fine in RyuJIT if `F` was consistently an 80-bit float. Then also `conv.r; nop; conv.r4` would give the same result. But without that it is hard to reconcile the behavior.

Of course any fix in RyuJIT also assumes that the final IL code is being compiled via RyuJIT. From a purely idealistic standpoint it's just not the right place to fix the problem if Roslyn wants to guarantee consistent behavior regardless of who ends up executing their IL. But I understand and agree about the pragmatism of this approach.

**Comment by tannergooding** [comment]
> I think it would be fine in RyuJIT if F was consistently an 80-bit float

I don't think we want to do this, strictly. It would be expensive to emulate for an edge case scenario that no one is going to care about or encounter in practice. Rather, I think we want to match what the actual expected behavior for IEEE 754 operations is, which is that any operation takes the inputs as given, then computes the result as if to infinite precision and unbounded range, and finally rounds to the nearest representable result.

That is, conceptually `F` should be "infinite precision and unbounded range". In practice we don't actually require that, however, because most inputs are going to be `F32` or `F64` and so for almost all real world code we have the following scenarios today:
* `F32 <op> F32 = F32`
* `F32 <op> F64 = F64`
* `F64 <op> F32 = F64`
* `F64 <op> F64 = F64`

Which is to say, the JIT correspondingly tracks the "proper" type in most cases and we know if its actually a `float` (F32) or `double` (F64). We don't allow doing things like `F32 + I32` either. The only exception to this, where we don't know the actual intended type, is `conv.r.un` where it takes a non-floating-point input and produces a generic `F` floating-point output of unspecified precision. We don't have any other instructions that operate like this, so this is a unique case that needs handling/consideration. That is all other instructions, despite saying they push `F` onto the stack, know the input types and precision they're operating in or specify an explicit target precision for the output, like `conv.r4/r8`.

Since the 3 main compilers always emit it so that `conv.r.un` is immediately followed by `conv.r4` or `conv.r8`, we don't really have any special considerations for typical code. In some edge cases, other tooling might insert some `nop` before the `conv.r4/r8`, which this would still consistently handle; but those should be atypical. Otherwise, its only manually emitted IL where we need to pick some behavior.

----------------------

So, I think we can keep our existing handling generally "as is". It has already been pushed towards IEEE 754 compliance, determinism, and we just need to specially treat the unique case of `conv.r.un`. We could then optionally define a new runtime feature flag (https://learn.microsoft.com/en-us/dotnet/api/system.runtime.compilerservices.runtimefeature?view=net-9.0) that tooling and languages could rely on to know if they should emit their own handling or if they can rely on what the runtime is doing.

To achieve this, we likely just want to ensure that `conv.r.un` for a `i64` or `u64` input produces some node that allows the JIT to know the target precision is indeterminate. It being consumed then "finalizes" that precision. If it is passed to `conv.r4/r8` we use that as the target precision. If it is spilled to a local, then it will have a target precision as per the local type (`F` is only valid on the evaluation stack). If it is passed to a method, we similarly know a target precision based on the type of the argument. So we then only have the unlikely case where its consumed by an operation like `add` (theoretically manual IL does `ldarg.0, ldarg.1, conv.r.un, add`) and where we should decide if it as the other input type such that `F <op> F32 = F32` or if we should treat it as the largest supported type `F <op> F32 = F64`. I don't think we'd want to have `F <op> F32 = F` due to the aforementioned expense in emulating arbitrary precision for an incredibly unlikely niche edge-case and would treat it as `F <op> F = F64` for simplicity. Given the unlikely nature of encountering this type of IL, I don't have a strong preference for which is chosen.

This should result in the least overall churn and still get us into the desired shape without needing to overly worry about increasingly unlikely edge cases.

**Comment by jakobbotsch** [comment]
There are a number of cases where we have to spill to JIT-created temporaries without a type that can be determined locally from IL, like e.g. because a value was left on the IL stack across control flow, or because side effects means we need to spill. So I also think this scheme would be quite complex to implement.
I would much prefer we either do the `conv.r.un; conv.r4` recognition and just accept this oddity in composability, or that we introduce a few library helpers for the unsigned conversions that IL producers can use starting in .NET 10.

**Comment by tannergooding** [comment]
> I would much prefer we either do the conv.r.un; conv.r4 recognition and just accept this oddity in composability

👍. This would be fine with me too.

> My preference is still on the latter, but if you prefer the former I am fine with it too.

Just to be clear, I'm not against some `RuntimeHelpers.ConvertToSingle` or similar and see the benefit of having some explicit API that guarantees it as well.

But I do think we need to ensure that `conv.r.un; conv.r4` is handled "correctly" (treated as a direct cast to `float`) given how the existing tooling have emitted the IL for 15+ years. I think we'd be much worse off if we ended up with netstandard2.0 libraries or other existing libraries having one behavior for `(float)someUInt64` and then only recompiled .NET 10+ libraries having a different behavior for the same code. It's one thing for .NET 9 vs .NET 10 to differ or .NET Framework vs .NET 10, but if the same library is differing based on whether it recompiled or not then we get into a much more complex range of breaking change considerations and bugs, especially for helper math libraries being consumed by applications or the like. I think accepting some oddities for unusual manual IL patterns is acceptable at that point and unlikely to be as broadly impacting to users.
